include/arch/arm/aarch64/arm_mmu.h

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2019 Broadcom
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_ARCH_ARM64_MMU_ARM_MMU_H_
+#define ZEPHYR_INCLUDE_ARCH_ARM64_MMU_ARM_MMU_H_
+
+/* Following Memory types supported through MAIR encodings can be passed
+ * by user through "attrs"(attributes) field of specified memory region.
+ * As MAIR supports such 8 encodings, we will reserve attrs[2:0];
+ * so that we can provide encodings upto 7 if needed in future.
+ */
+#define MT_TYPE_MASK		0x7U
+#define MT_TYPE(attr)		(attr & MT_TYPE_MASK)
+#define MT_DEVICE_nGnRnE	0U
+#define MT_DEVICE_nGnRE		1U
+#define MT_DEVICE_GRE		2U
+#define MT_NORMAL_NC		3U
+#define MT_NORMAL		4U
+
+#define MEMORY_ATTRIBUTES	((0x00 << (MT_DEVICE_nGnRnE * 8)) |	\
+				(0x04 << (MT_DEVICE_nGnRE * 8))   |	\
+				(0x0c << (MT_DEVICE_GRE * 8))     |	\
+				(0x44 << (MT_NORMAL_NC * 8))      |	\
+				(0xffUL << (MT_NORMAL * 8)))
+
+/* More flags from user's perpective are supported using remaining bits
+ * of "attrs" field, i.e. attrs[31:3], underlying code will take care
+ * of setting PTE fields correctly.
+ *
+ * current usage of attrs[31:3] is:
+ * attrs[3] : Access Permissions
+ * attrs[4] : Memory access from secure/ns state
+ * attrs[5] : Execute Permissions
+ *
+ */
+#define MT_PERM_SHIFT		3U
+#define MT_SEC_SHIFT		4U
+#define MT_EXECUTE_SHIFT	5U
+
+#define MT_RO			(0U << MT_PERM_SHIFT)
+#define MT_RW			(1U << MT_PERM_SHIFT)
+
+#define MT_SECURE		(0U << MT_SEC_SHIFT)
+#define MT_NS			(1U << MT_SEC_SHIFT)
+
+#define MT_EXECUTE		(0U << MT_EXECUTE_SHIFT)
+#define MT_EXECUTE_NEVER	(1U << MT_EXECUTE_SHIFT)
+
+/* Some compound attributes for most common usages */
+#define MT_CODE			(MT_NORMAL | MT_RO | MT_EXECUTE)
+#define MT_RODATA		(MT_NORMAL | MT_RO | MT_EXECUTE_NEVER)
+
+/*
+ * PTE descriptor can be Block descriptor or Table descriptor
+ * or Page descriptor.
+ */
+#define PTE_DESC_TYPE_MASK	3U
+#define PTE_BLOCK_DESC		1U
+#define PTE_TABLE_DESC		3U
+#define PTE_PAGE_DESC		3U
+#define PTE_INVALID_DESC	0U
+
+/*
+ * Block and Page descriptor attributes fields
+ */
+#define PTE_BLOCK_DESC_MEMTYPE(x)	(x << 2)
+#define PTE_BLOCK_DESC_NS		(1UL << 5)
+#define PTE_BLOCK_DESC_AP_RO		(1UL << 7)
+#define PTE_BLOCK_DESC_AP_RW		(0UL << 7)
+#define PTE_BLOCK_DESC_NON_SHARE	(0UL << 8)
+#define PTE_BLOCK_DESC_OUTER_SHARE	(2UL << 8)
+#define PTE_BLOCK_DESC_INNER_SHARE	(3UL << 8)
+#define PTE_BLOCK_DESC_AF		(1UL << 10)
+#define PTE_BLOCK_DESC_NG		(1UL << 11)
+#define PTE_BLOCK_DESC_PXN		(1UL << 53)
+#define PTE_BLOCK_DESC_UXN		(1UL << 54)
```

Comment:
Nit: `ULL` would be a better choice (and aligns with what `TCR` definitions are doing).